activesupport/lib/active_support/duration/iso8601_parser.rb

```diff
@@ -0,0 +1,72 @@
+require 'active_support/core_ext/hash/slice'
+
+module ActiveSupport
+  class Duration
+    # Parses a string formatted according to ISO 8601 Duration into the hash .
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601Parser
+      attr_reader :parts
+
+      class ParsingError < ::StandardError; end
+
+      def initialize(iso8601duration)
+        @raw = iso8601duration
+        parse_iso_duration!
+        construct_parts
+        validate_parts!
+      end
+
+      private
+        def parse_iso_duration!
+          @match = @raw.match(/^
+                      (?<sign>\+|-)?
```

Comment:
This PR is a basis for another PR #16919 to work with interval datatype of PostgreSQL, and Postgres docs provides examples of such scary durations, so we should support them. See http://www.postgresql.org/docs/9.4/static/datatype-datetime.html#DATATYPE-INTERVAL-OUTPUT at most bottom. Postgres usually well known for it's strictness about everything.